tensorboard/plugins/graph/README.md

```diff
@@ -121,9 +121,384 @@ describing the motivations, research, and process behind it.
+components.
+
+Note: building a Hierarchy modifies the SlimGraph's OpNodes in place!
+
+**tf_graph_common/hierarchy.ts: build(SlimGraph): Hierarchy**
+
+-   Create an artificial MetaNode for the root.
+-   Create an index of nodes as `Record<nodeId: string,
+    OpNode|SeriesNode|MetaNode>`
+-   'Adding nodes'
+    -   Iterate over OpNodes, keep track of device names, XLAClusters,
+        opHistograms, TPU Compatibility histograms, library function nodes and
+        their associated function name, etc.
+    -   Assume the node's name's '/' occurences indicate its level from the
+        root. If node is named 'a/b/c', create metanodes 'a' and 'a/b'.
+    -   Update embedding nodes to point to the OpNode they are embedded to:
+        `embedding.parentNode = node;`.
+-   'Detect series'
+    -   Recursively traverse metanodes starting from the top-level. At each
+        metanode,
+        -   Take the direct children nodes and group them into clusters by
+            op-type.
+        -   For each op cluster, detect series by greedily looking for numeric
+            suffix changes. For example, 3 OpNodes with op 'Add': `add_1`,
+            `add_2`, `add_3` will be grouped into a SeriesNode 'add_[1-3]'.
+        -   Series groups require at least 2 nodes (later we filter out groups
+            with less than 5 nodes) where the numeric suffix is in increasing
+            order.
+        -   Children within a SeriesNode have a reference: `child.owningSeries =
+            seriesName`.
+    -   If series has 5+ members,
+        -   Add it to hierarchy and the metagraph, updating references,
+            histograms, etc. The metagraph no longer considers the original
+            OpNodes as direct children, if they are owned by a SeriesNode.
+-   'Adding edges'
+    -   For each BaseEdge
+        -   Instead of creating an edge between the two nodes, create a MetaEdge
+            between their ancestors underneath their common ancestor furthest
+            from the root. Update the shared ancestor MetaNode's metagraph.
+        -   For each metanode with a metaedge, store whether it has a 'non
+            control dependency' edge.
+        -   Compute the size of the tensor flowing through this base edge and
+            store the maxMetaEdgeSize on the Hierarchy itself.
+            -   For each outputShape tensor, get the "size": [4, 2, 1] is 4 *
+                2 * 1 = 12.
+            -   When a metaedge contains multiple base edges, it sums their
+                total sizes
+        -   Update the MetaEdge with a reference to the BaseEdge:
+            `metaedge.addBaseEdge(baseEdge, h)`. A metanode may contain multiple
+            BaseEdges if there are multiple connections between two metanodes.
+    -   Note, the Hierarchy does not have a direct reference to metaedges. It
+        simply has a root metanode with a metagraph, and edges of the metagraph
```

Comment:
Hmm, so who owns the metaedge? Is it metanode?

Reply:
Good call, it's the metagraph.  Updated.